packages/common/src/types/Module.ts

```diff
@@ -0,0 +1,119 @@
+import { ModuleReference } from './moduleReference';
+import * as H from '../contractHelpers';
+import { sha256 } from '../hash';
+import { Buffer } from 'buffer/';
+import { VersionedModuleSource } from '../types';
+
+/** Interface of a smart contract containing the name of the contract and every entrypoint. */
+export type ContractInterface = {
+    /** The name of the smart contract. Note: This does _not_ including the 'init_' prefix. */
+    contractName: H.ContractName;
+    /** A set of entrypoints exposed by the smart contract. Note: These does _not_ include the '<contractName>.' prefix. */
+    entrypointNames: Set<H.EntrypointName>;
+};
+
+/** Interface of a smart contract module containing the interface of every contract in the module. */
+export type ModuleInterface = Map<H.ContractName, ContractInterface>;
+
+/**
+ * A versioned smart contract module.
+ */
+export class Module {
+    /** Cache of the parsed WebAssembly module. */
+    private wasmModule: WebAssembly.Module | undefined;
+    /** Cache of the module reference. */
+    private moduleRef: ModuleReference | undefined;
+
+    private constructor(
+        /** The version of the smart contract module. */
+        public version: number,
+        /** Bytes for the WebAssembly module. */
+        public moduleSource: Buffer
+    ) {}
+
+    /**
+     * Construct a smart contract module object from bytes, potentially read from a file.
+     * @param bytes Bytes encoding a versioned smart contract module.
+     */
+    public static fromRawBytes(bytes: Buffer): Module {
+        const version = bytes.readInt32LE(0);
+        const moduleSource = bytes.subarray(8);
+        return new Module(version, Buffer.from(moduleSource));
+    }
+
+    /**
+     * Contruct a smart contract module object from a versioned module source.
+     * @param versionedModule The versioned module.
+     */
+    public static fromModuleSource(
+        versionedModule: VersionedModuleSource
+    ): Module {
+        return new Module(versionedModule.version, versionedModule.source);
+    }
+
+    /** Calculate the module reference from the module source. The module reference is cached. */
+    public getModuleRef(): ModuleReference {
+        if (this.moduleRef !== undefined) {
+            return this.moduleRef;
+        }
+        const hash = sha256([this.moduleSource]);
```

Comment:
Are you sure this is correct. In Rust we have
```
    pub fn get_module_ref(&self) -> ModuleReference {
        let mut hasher = sha2::Sha256::new();
        self.serial(&mut hasher);
        ModuleReference::from(<[u8; 32]>::from(hasher.finalize()))
    }
```

so what is hashed is
- version
- length prefix (4 bytes, big endian)
- the actual source